Support min_by/max_by group by aggregate

thirtiseven opened the pull request:
## Description

Closes https://github.com/rapidsai/cudf/issues/16139

This pr adds support for min_by, which is used to return the value of a column associated with the minimum value of another column. It will be useful for spark-rapids.

Currently this pr only supports sort based group by, will try to add a hash group by too, but I'm not very clear how to do it right now because the input column from spark will be a struct column of value and order.

Related pr in spark-rapids: https://github.com/NVIDIA/spark-rapids/pull/11123

For Spark, all orderable types (basic types and array/struct) are supported, except float and double with NaN values, because Spark has a special handling for NaN in non-nested floating types.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

bdice reviewed (changes requested):
Can you explain an example of where this new aggregation is a better choice than argmin + gather? libcudf already provides the essential building blocks for this kind of operation, and I don't see how this specialized implementation provides a significant benefit. I'm not sure if I find the claims in #16139 about memory pressure, single-pass aggregation, and performance to be compelling from a surface level view.

If we do go this route and decide this feature is necessary, we should also implement a `max_by` aggregation at the same time, for symmetry.

thirtiseven commented:
> Can you explain an example of where this new aggregation is a better choice than argmin + gather? libcudf already provides the essential building blocks for this kind of operation, and I don't see how this specialized implementation provides a significant benefit. I'm not sure if I find the claims in #16139 about memory pressure, single-pass aggregation, and performance to be compelling from a surface level view.

min_by is an aggregation for Spark but not in Pandas, so we would like to match Spark's behavior directly from the cuDF side. argmin/max + gather have few features gap between Spark's min/max_by that are difficult to handle from the spark-rapids side, such as
- all nulls in a grouped order column, Spark returns null, argmin/max + gather returns first element in grouped value column.
- NaN for float aggregation. In Spark, `Nan` is the maximum float value, but in cuDF, the calculation involving `Nan` is undefined.
- min_by and max_by return the last minimum order value in Spark. For argmin it's matched, but for argmax cuDF will return the first maximum order value.

Now we have handled the all nulls case by modifying the null masks and using argmin+gather route to quickly support our customer's need. But for the next step, we'd like to have an independent implementation to support float aggregation and max_by.

Another reason is that min_by and max_by are two special aggregations in Spark that need to perform aggregation on two different columns. So we need to package the two columns into one struct column for cuDF to handle because AFAIK cuDF only supports aggregation on one column. Otherwise we need to do some special post-processing in spark-rapids to check if there are min/max_bys in aggs, gather them to their value column and concat the results back into the original agg result table, which would be much slower and harder to implement. 

> If we do go this route and decide this feature is necessary, we should also implement a `max_by` aggregation at the same time, for symmetry.

Will do if we go this route.

bdice commented:
In terms of semantics, the proposed min_by would need to match the argmin + gather implementation exactly. In libcudf, null and NaN control are handled by separate arguments. For example, the collect set aggregation (https://docs.rapids.ai/api/libcudf/stable/group__aggregation__factories#gaebe680a414f3c942a631f609bcfb5781) accepts null_policy, null_equality, and nan_equality arguments. This is a better route to address the desired semantics if you can express it. There are also some examples of ordering policies in libcudf, like null_order.

We need to work on some improvements to argmin and argmax anyway, so this would be a good joint project for us. Currently we have “experimental” row comparators that we use for everything except for argmin calls, and we need to adopt those to expand support for nested list and struct columns. I can dig up the issue where this is discussed.

bdice commented:
@thirtiseven Here is the other issue I was thinking of: https://github.com/rapidsai/cudf/issues/14412#issuecomment-1811231902

thirtiseven commented:
> In terms of semantics, the proposed min_by would need to match the argmin + gather implementation exactly. In libcudf, null and NaN control are handled by separate arguments. For example, the collect set aggregation (https://docs.rapids.ai/api/libcudf/stable/group__aggregation__factories#gaebe680a414f3c942a631f609bcfb5781) accepts null_policy, null_equality, and nan_equality arguments. This is a better route to address the desired semantics if you can express it. There are also some examples of ordering policies in libcudf, like null_order.

That sounds good, adding those arguments seems to work. I will try to do this, maybe in another pr just for argmin/argmax and make min/max_by just: 1. unpack the struct column. 2. call argmin/argmax with different arguments. 3. do a gather with the struct column. Does this make sense to you?

> We need to work on some improvements to argmin and argmax anyway, so this would be a good joint project for us. Currently we have “experimental” row comparators that we use for everything except for argmin calls, and we need to adopt those to expand support for nested list and struct columns. I can dig up the issue where this is discussed.
> @thirtiseven Here is the other issue I was thinking of: [#14412 (comment)](https://github.com/rapidsai/cudf/issues/14412#issuecomment-1811231902)

Thank you! I was wondering how to write a hash-based implementation for min_by since the value will always be a struct (but no idea). 
Currently for spark-rapids the sort-based way is good enough according to perf tests. We'd love to write a sort-based min/max_by as the next step after argmin/max supports nested types with row comparators.

thirtiseven commented:
Updated code to added support for max_by and removed group_min_by.cu to use argmin + gather.

There is a small post-processing to replace all nulls to -1 after calling argmin. In argmin's code, it says indices corresponding to null values was initialized to ARGMIN_SENTINEL. but it is not initialized in current code. I think it's not a bug, but some outdated comments, but I don't have much context about it. Filed an issue #16542 about it.

Please take another look thanks @bdice

ttnghia commented on cpp/include/cudf/aggregation.hpp:
```diff
@@ -120,7 +120,9 @@ class aggregation {
     HISTOGRAM,       ///< compute frequency of each element
-    MERGE_HISTOGRAM  ///< merge partial values of HISTOGRAM aggregation,
+    MERGE_HISTOGRAM, ///< merge partial values of HISTOGRAM aggregation,
+    MAX_BY,          ///< max reduction by another column
+    MIN_BY           ///< min reduction by another column
```
Since we are specifying just one `value` column (which is a struct of 2 columns), this comment is not accurate. It should be s something like `retrieve value on a child column based on min/max value of another child column`.

thirtiseven replied:
updated.

ttnghia commented on cpp/include/cudf/aggregation.hpp:
```diff
@@ -382,6 +384,28 @@ std::unique_ptr<Base> make_argmax_aggregation();
+/**
+ * @brief Factory to create a MAX_BY aggregation
+ *
+ * `MAX_BY` returns the value of the element in the group that is the minimum
+ * according to the order_by column.
```
What is the `order_by` column here?
```suggestion
 * `MAX_BY` returns values of the first child column that correspond to the minimum values of the second child column.
```

thirtiseven replied:
updated.

ttnghia commented on cpp/include/cudf/detail/aggregation/aggregation.hpp:
```diff
@@ -1219,6 +1261,18 @@ struct target_type_impl<Source, aggregation::MIN> {
 };
 
+// Computing MAX_BY of Source, use Source accumulator
+template <typename Source>
+struct target_type_impl<Source, aggregation::MAX_BY> {
+  using type = struct_view;
+};
+
+// Computing MIN_BY of Source, use Source accumulator
+template <typename Source>
+struct target_type_impl<Source, aggregation::MIN_BY> {
+  using type = struct_view;
```
So the output type is the same type as the input? If so, I would recommend to add a parameter to the constructor of min_by/max_by aggregation to specify which column index to compute arg_min/arg_max. By doing so, if in the future we need to compute min_by/max_by of more than one column (i.e., get values of several columns based on the same min/max value of the order by column) then we will trivially have this supported.

thirtiseven replied:
> So the output type is the same type as the input?

Yes.

The struct is packed manually in plugin when doing min/max_by, not the original schema of table. So we can just always make the column which need to be argmin/max as the last column without specifying the index.

thirtiseven replied:
Oh, current code actually always selects the second column. Will make the ordering column be the first one and noted in comment.

thirtiseven replied:
done.

ttnghia replied:
> the ordering column be the first one

Yes the first column as the ordering column makes more sense than the last one :+1:

ttnghia reviewed (changes requested):
Recommend to change the constructors of the aggregations so we can support batch `min_by`/`max_by`.

thirtiseven commented on cpp/include/cudf/aggregation.hpp:
```diff
@@ -382,6 +384,28 @@ std::unique_ptr<Base> make_argmax_aggregation();
+ * @brief Factory to create a MIN_BY aggregation
+ *
+ * `MIN_BY` returns the value of the element in the group that is the minimum
+ * according to the order_by column.
```
updated.

thirtiseven commented on cpp/src/groupby/sort/aggregate.cpp:
```diff
@@ -198,6 +199,67 @@ void aggregate_result_functor::operator()<aggregation::MIN>(aggregation const& a
+{
+  if (cache.has_result(values, agg)) return;
+
+  auto argmax_result = detail::group_argmax(get_grouped_values().child(1),
```
done.

ttnghia commented on cpp/include/cudf/aggregation.hpp:
```diff
@@ -120,7 +120,9 @@ class aggregation {
     HISTOGRAM,       ///< compute frequency of each element
-    MERGE_HISTOGRAM  ///< merge partial values of HISTOGRAM aggregation,
+    MERGE_HISTOGRAM, ///< merge partial values of HISTOGRAM aggregation,
+    MAX_BY,          ///< retrieve values based on max of another column
+    MIN_BY           ///< retrieve values based on min of another column
```
```suggestion
    MAX_BY,          ///< retrieve rows based on max values of the first child column
    MIN_BY           ///< retrieve rows based on min values of the first child column
```

ttnghia commented on cpp/src/groupby/sort/aggregate.cpp:
```diff
@@ -198,6 +199,71 @@ void aggregate_result_functor::operator()<aggregation::MIN>(aggregation const& a
+{
+  if (cache.has_result(values, agg)) return;
+
+  CUDF_EXPECTS(values.num_children() >= 2, "Input column must have at least 2 children.");
```
```suggestion
  CUDF_EXPECTS(values.type().id() == type_id::STRUCT && values.num_children() >= 2, "Input column must have at least 2 children.");
```

ttnghia commented on cpp/src/groupby/sort/aggregate.cpp:
```diff
@@ -198,6 +199,71 @@ void aggregate_result_functor::operator()<aggregation::MIN>(aggregation const& a
+
+  CUDF_EXPECTS(values.num_children() >= 2, "Input column must have at least 2 children.");
+
+  auto argmax_result = detail::group_argmax(get_grouped_values().child(0),
```
```suggestion
  auto const argmax_result = detail::group_argmax(get_grouped_values().child(0),
```

ttnghia commented on cpp/src/groupby/sort/aggregate.cpp:
```diff
@@ -198,6 +199,71 @@ void aggregate_result_functor::operator()<aggregation::MIN>(aggregation const& a
+                                            stream,
+                                            mr);
+
+  auto sentinel = cudf::numeric_scalar<size_type>{cudf::detail::ARGMAX_SENTINEL};
```
```suggestion
  auto const sentinel = cudf::numeric_scalar<size_type>{cudf::detail::ARGMAX_SENTINEL};
```

ttnghia commented on cpp/src/groupby/sort/aggregate.cpp:
```diff
@@ -198,6 +199,71 @@ void aggregate_result_functor::operator()<aggregation::MIN>(aggregation const& a
+                                            mr);
+
+  auto sentinel = cudf::numeric_scalar<size_type>{cudf::detail::ARGMAX_SENTINEL};
+
+  auto null_removed_map = cudf::replace_nulls(
```
```suggestion
  auto const null_removed_map = cudf::replace_nulls(
```

ttnghia replied:
```suggestion
  auto const null_removed_map = cudf::replace_nulls(
```

ttnghia commented on cpp/src/groupby/sort/aggregate.cpp:
```diff
@@ -198,6 +199,71 @@ void aggregate_result_functor::operator()<aggregation::MIN>(aggregation const& a
+{
+  if (cache.has_result(values, agg)) return;
+
+  CUDF_EXPECTS(values.num_children() >= 2, "Input column must have at least 2 children.");
```
```suggestion
  CUDF_EXPECTS(values.type().id() == type_id::STRUCT && values.num_children() >= 2, "Input column must have at least 2 children.");
```

ttnghia commented on cpp/src/groupby/sort/aggregate.cpp:
```diff
@@ -198,6 +199,71 @@ void aggregate_result_functor::operator()<aggregation::MIN>(aggregation const& a
+
+  CUDF_EXPECTS(values.num_children() >= 2, "Input column must have at least 2 children.");
+
+  auto argmin_result = detail::group_argmin(get_grouped_values().child(0),
+                                            helper.num_groups(stream),
+                                            helper.group_labels(stream),
+                                            helper.key_sort_order(stream),
+                                            stream,
+                                            mr);
+
+  auto sentinel = cudf::numeric_scalar<size_type>{cudf::detail::ARGMIN_SENTINEL};
+
+  auto null_removed_map = cudf::replace_nulls(
+    argmin_result->view(), sentinel, stream, mr);
```
```suggestion
  auto const argmin_result = detail::group_argmin(get_grouped_values().child(0),
                                            helper.num_groups(stream),
                                            helper.group_labels(stream),
                                            helper.key_sort_order(stream),
                                            stream,
                                            mr);

  auto const sentinel = cudf::numeric_scalar<size_type>{cudf::detail::ARGMIN_SENTINEL};
  auto const null_removed_map = cudf::replace_nulls(
    argmin_result->view(), sentinel, stream, mr);
```

ttnghia commented on cpp/src/groupby/sort/group_reductions.hpp:
```diff
@@ -100,6 +100,53 @@ std::unique_ptr<column> group_min(column_view const& values,
                                   rmm::cuda_stream_view stream,
                                   rmm::device_async_resource_ref mr);
+/**
+ * @brief Internal API to calculate groupwise maximum value by another order column, packed in a
+ * structs column.
```
```suggestion
 * @brief Internal API to find rows corresponding to the groupwise maximum value of the first child in the values column.
 *
 * The values column is always given as a structs column having at least two children.
```

ttnghia commented on cpp/src/groupby/sort/group_reductions.hpp:
```diff
@@ -100,6 +100,53 @@ std::unique_ptr<column> group_min(column_view const& values,
+ * group_max_by = [{4, 4}, {-1, 3}, {4, 4}, {<NA>, 2}]
+ * @endcode
+ * 
+ * @param structs_column Structs column containing the values to get maximum from
+ * @param group_labels ID of group that the corresponding value belongs to
+ * @param num_groups Number of groups
+ * @param stream CUDA stream used for device memory operations and kernel launches.
+ * @param mr Device memory resource used to allocate the returned column's device memory
+ */
+std::unique_ptr<column> group_max_by(column_view const& structs_column,
```
```suggestion
 * @param values The input structs column
 * @param group_labels ID of groups that the corresponding value belongs to
 * @param num_groups Number of groups
 * @param stream CUDA stream used for device memory operations and kernel launches.
 * @param mr Device memory resource used to allocate the returned column's device memory
 */
std::unique_ptr<column> group_max_by(column_view const& values,
```

ttnghia commented on cpp/src/groupby/sort/group_reductions.hpp:
```diff
@@ -100,6 +100,53 @@ std::unique_ptr<column> group_min(column_view const& values,
+                                     rmm::device_async_resource_ref mr);
+
+/**
+ * @brief Internal API to calculate groupwise minimum value by another order column, packed in a
+ * structs column.
```
Same suggestion for this function as for `max_by`.

ttnghia commented on cpp/tests/groupby/argmin_tests.cpp:
```diff
@@ -91,6 +91,26 @@ TYPED_TEST(groupby_argmin_test, zero_valid_values)
   test_single_agg(keys, vals, expect_keys, expect_vals, std::move(agg2), force_use_sort_impl::YES);
 }
 
+TYPED_TEST(groupby_argmin_test, zero_valid_values_xx)
```
Please create a separate test such as `groupby_minby_test`, do not use existing test for easier debugging.

ttnghia replied:
Why do we have this test?

ttnghia commented on cpp/tests/groupby/max_by_tests.cpp:
```diff
@@ -0,0 +1,77 @@
+
+  cudf::test::fixed_width_column_wrapper<K> keys{1, 2, 3, 1, 2, 2, 1, 3, 3, 2};
+  cudf::test::fixed_width_column_wrapper<V> orders{1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
+  cudf::test::fixed_width_column_wrapper<K> values{4, 1, 2, 3, 4, 5, 6, 7, 8, 9};
+  cudf::test::structs_column_wrapper vals{orders, values};
```
We should better call the second column `payload` instead of values, since the entire structs column is input `values` into groupby.

ttnghia replied:
(This applies for all new added tests)

ttnghia commented on java/rmm_log.txt:
```diff
@@ -0,0 +1,10 @@
+[3231126][17:00:30:499434][info  ] ----- RMM LOG BEGIN [PTDS DISABLED] -----
```
Added by mistake?

davidwendt commented:
> Updated code to added support for max_by and removed group_min_by.cu to use argmin + gather.
> 
> There is a small post-processing to replace all nulls to -1 after calling argmin. In argmin's code, it says indices corresponding to null values was initialized to ARGMIN_SENTINEL. but it is not initialized in current code. I think it's not a bug, but some outdated comments, but I don't have much context about it. Filed an issue #16542 about it.
> 
> Please take another look thanks @bdice

Seems like this aggregation is just calling other libcudf APIs: `groupby(argmax)`, `replace_nulls`, and `gather`.
I'm not convinced this needs to be implemented by libcudf or is providing any specific value.
The caller can just make these 3 calls. I recommend closing this PR.
